[deepspeed doc] install issues + 1-gpu deployment (#9582)
huggingface/transformers: docs/source/main_classes/trainer.rst

This provided support is new and experimental as of this writing.

Installation Notes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

As of this writing, both, FairScale and Deepspeed, require compilation of CUDA C++ code, before they can be used.

> While all installation issues should be dealt with through the corresponding github Issues of FairScale and DeepSpeed,

Reviewer: ```suggestion
While all installation issues should be dealt with through the corresponding GitHub Issues of FairScale and DeepSpeed,
```

Reviewer: + links to issues of FairScale and DeepSpeed (under their respective names) here?

> there are a few common issues that one may encounter while building any pytorch extension that needs to build CUDA

Reviewer: ```suggestion
there are a few common issues that one may encounter while building any PyTorch extension that needs to build CUDA
```

extensions.

Therefore, if you encounter a CUDA-related build issue while doing one of the following or both:

.. code-block:: bash

    pip install fairscale
    pip install deepspeed

please, read the following notes first.

In these notes we give examples for what to do when ``pytorch`` has been built with CUDA ``10.2``. If your situation is
different remember to adjust the version number to the one you are after.

**Possible problem #1:**

> While, pytorch comes with its own CUDA toolkit, to build these two projects you must have an identical version of CUDA

Reviewer: ```suggestion
While, PyTorch comes with its own CUDA toolkit, to build these two projects you must have an identical version of CUDA
```
(Everywhere in text pytorch -> PyTorch, the \`\`pytorch\`\` in backsticks can stay the same, will stop highlighting)

installed system-wide.

> For example, if you installed ``pytorch`` with ``cudatoolkit==10.2`` in the python environment, you also need to have

Reviewer: ```suggestion
For example, if you installed ``pytorch`` with ``cudatoolkit==10.2`` in the Python environment, you also need to have
```
Everywhere python -> Python while we are at it :-)

CUDA ``10.2`` installed system-wide.

The exact location may vary from system to system, but ``/usr/local/cuda-10.2`` is the most common location on many
Unix systems. If you don't have it installed, install it first. You will find the instructions by using your favorite
search engine. For example, if you're on Ubuntu you may want to search for: `ubuntu cuda 10.2 install

> <https://www.google.com/search?q=ubuntu+cuda+10.2+install>`__.

Author: great tip - added!

**Possible problem #3:**

Some older CUDA versions may refuse to build with newer compilers. For example, you my have ``gcc-9`` but it wants
``gcc-7``.

There are various ways to go about it.

If you can install the latest CUDA toolkit it typically should support the newer compiler.

Alternatively, you could install the lower version of the compiler in addition to the one you already have, or you may
already have it but it's not the default one, so the build system can't see it. If you have ``gcc-7`` installed but the
build system complains it can't find it, the following might do the trick:

.. code-block:: bash

    sudo ln -s /usr/bin/gcc-7  /usr/local/cuda-10.2/bin/gcc
    sudo ln -s /usr/bin/g++-7  /usr/local/cuda-10.2/bin/g++

Here, we are making a symlink to ``gcc-7`` from ``/usr/local/cuda-10.2/bin/gcc`` and since
``/usr/local/cuda-10.2/bin/`` should be in the ``PATH`` environment variable (see the previous problem's solution), it
should find ``gcc-7`` (and ``g++7``) and then the build will succeed.

As always make sure to edit the paths in the example to match your situation.

**If still unsuccessful:**

> If after addressing these you still encounter build issues, please, proceed with the github Issue of `FairScale

Reviewer: ```suggestion
If after addressing these you still encounter build issues, please, proceed with the GitHub Issue of `FairScale
```

1. Optimizer State Sharding
2. Gradient Sharding

> You will need at least 2 GPUs to use this feature.

Reviewer: ```suggestion
You will need at least two GPUs to use this feature.
```

DeepSpeed
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

`DeepSpeed <https://github.com/microsoft/DeepSpeed>`__ implements everything described in the `ZeRO paper
<https://arxiv.org/abs/1910.02054>`__, except ZeRO's stage 3. "Parameter Partitioning (Pos+g+p)". Currently it provides
full support for:

1. Optimizer State Partitioning (ZeRO stage 1)
2. Add Gradient Partitioning (ZeRO stage 2)

Installation
=======================================================================================================================

Install the library via pypi:

.. code-block:: bash

    pip install deepspeed

> or find more details on `the DeepSpeed's github page <https://github.com/microsoft/deepspeed#installation>`__.

Reviewer: ```suggestion
or find more details on `the DeepSpeed's GitHub page <https://github.com/microsoft/deepspeed#installation>`__.
```
Everywhere github -> GitHub ;-)

full details on how to configure various nodes and GPUs can be found `here
<https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node>`__.

Here is an example of running ``finetune_trainer.py`` under DeepSpeed deploying all available GPUs:

.. code-block:: bash

    cd examples/seq2seq
    deepspeed ./finetune_trainer.py --deepspeed ds_config.json \
    --model_name_or_path sshleifer/distill-mbart-en-ro-12-4 --data_dir wmt_en_ro \
    --output_dir output_dir --overwrite_output_dir \
    --do_train --n_train 500 --num_train_epochs 1 \
    --per_device_train_batch_size 1  --freeze_embeds \
    --src_lang en_XX --tgt_lang ro_RO --task translation

Note that in the DeepSpeed documentation you are likely to see ``--deepspeed --deepspeed_config ds_config.json`` - i.e.
two DeepSpeed-related arguments, but for the sake of simplicity, and since there are already so many arguments to deal
with, we combined the two into a single argument.

For some practical usage examples, please, see this `post
<https://github.com/huggingface/transformers/issues/8771#issuecomment-759248400>`__.

> Deployment with 1 GPU

Reviewer: ```suggestion
Deployment with one GPU
```

=======================================================================================================================

> To deploy DeepSpeed with 1 GPUs adjust the :class:`~transformers.Trainer` command line arguments as following:

Reviewer: ```suggestion
To deploy DeepSpeed with one GPU adjust the :class:`~transformers.Trainer` command line arguments as following:
```

.. code-block:: bash

    cd examples/seq2seq
    deepspeed --num_gpus=1 ./finetune_trainer.py --deepspeed ds_config.json \
    --model_name_or_path sshleifer/distill-mbart-en-ro-12-4 --data_dir wmt_en_ro \
    --output_dir output_dir --overwrite_output_dir \
    --do_train --n_train 500 --num_train_epochs 1 \
    --per_device_train_batch_size 1  --freeze_embeds \
    --src_lang en_XX --tgt_lang ro_RO --task translation

This is almost the same as with multiple-GPUs, but here we tell DeepSpeed explicitly to use just one GPU. By default,
DeepSpeed deploys all GPUs it can see. If you have only 1 GPU to start with, then you don't need this argument. The
following `documentation <https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node>`__ discusses the
launcher options.

> Why would you want to use DeepSpeed with just 1 GPU?

Reviewer: ```suggestion
Why would you want to use DeepSpeed with just one GPU?
```
